### DISCUSSION TITLE
Control the size of the S3 bucket used for WAL archiving

### DISCUSSION THREAD

**Discussion by pietrobaricco** [cloudnative-pg/cloudnative-pg#4051] [Q&A]
Hello, I've replaced a number of crunchy PGO clusters I had with cnpg.
All is well except for one thing: the size of the S3 buckets used for WAL archiving keeps growing indefinitely.
I'd like to have a 10-15 days of PITR window, and then rely on snapshots for older backups but haven't found any clear indications on how to achieve this in the docs.
Thanks!

**Comment by gbartolini** [comment]
CloudNativePG has no direct visibility on the object stores, as it relies on Barman Cloud. Our direction is to make CNPG backup-agnostic and delegate this to plugins.

Regarding snapshots, it is your choice. You can use multiple backup strategies and, for example, have scheduled backups on volume snapshots to be taken once a month and be offline - therefore consistent.

**Comment by pietrobaricco** [comment]
FYI I had to implement myself a contraption which

- queries the S3 bucket determining the oldest backup to keep (can be parametrized by date or nr of backups to retain)
- deletes all backups older than said date
- deletes all WALS older than said date

It works and finally made my S3 storage consumption manageable, as there were several terabytes of cruft there doing literally nothing from the beginning of times.
Another thing I noticed is that, even if a certain backup gets deleted via the cnpg krew plugin, it is not physically deleted from S3.

S3 storage costs money, and in some cases there is a ceiling on how much data can be stored. If nothing ever gets deleted, at a certain point no more WALs or backups can be stored and all kinds of nastiness can and will occur.
I totally get the "not my problem" attitude, just be mindful that for many people less willing than me to roll their own cleanup procedures, this issue will be a deal breaker, or will leave a very sour taste when shtf.
